src/main/java/ome/services/graphs/GraphTraversal.java

```diff
@@ -1213,6 +1215,53 @@ private void reviewObject(CI object, boolean isErrorRules) throws GraphException
         }
     }
 
+    /**
+     * Note the directory contents that must be deleted before the parent directory is deleted.
+     */
+    private void orderFileDeletion() {
+        /* find which files are to be deleted */
+        final String originalFileClassName = OriginalFile.class.getName();
+        final Set<Long> originalFileIds = new HashSet<>();
+        for (final CI object : planning.deleted) {
+            if (originalFileClassName.equals(object.className)) {
+                originalFileIds.add(object.id);
+            }
+        }
+        /* find which of those files are directories in a repository */
+        final SetMultimap<String, Long> directoryRepos = HashMultimap.create();
+        final String directoryRepoQuery =
+                "SELECT id, repo FROM OriginalFile WHERE mimetype = 'Directory' AND repo IS NOT NULL AND id IN (:ids)";
+        for (final List<Long> originalFileIdsBatch : Iterables.partition(originalFileIds, BATCH_SIZE)) {
+            for (final Object[] result :
+                (List<Object[]>) session.createQuery(directoryRepoQuery).setParameterList("ids", originalFileIdsBatch).list()) {
```

Comment:
This was quite hard to read. It would be better to split it out to a variable, or better yet a function as there is already a number of variables hanging around.